lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2467,6 +2467,170 @@ static bool checkFunctionSignature(
+            ? selfType
+            : selfType->mapTypeOutOfContext();
+        selfType =
+            derivativeGenEnv->mapTypeIntoContext(selfInterfaceType);
+      }
+      if (!conformsToDifferentiableInModule(
+              selfType, function->getModuleContext())) {
+        TC.diagnose(attrLoc, diag::diff_function_no_parameters,
+                    function->getFullName())
+            .highlight(function->getSignatureSourceRange());
+        return nullptr;
+      }
+    }
+  }
+
+  // If parsed differentiation parameters are empty, infer parameter indices
+  // from the function type.
+  if (parsedWrtParams.empty()) {
+    return AutoDiffParameterIndicesBuilder::inferParameters(
+        functionType, function->getModuleContext())
+        .build(TC.Context);
+  }
+
+  // Otherwise, build parameter indices from parsed differentiation parameters.
+  AutoDiffParameterIndicesBuilder builder(functionType);
+  int lastIndex = -1;
+  for (unsigned i : indices(parsedWrtParams)) {
+    auto paramLoc = parsedWrtParams[i].getLoc();
+    switch (parsedWrtParams[i].getKind()) {
+      case ParsedAutoDiffParameter::Kind::Named: {
+        auto nameIter =
+            llvm::find_if(params.getArray(), [&](ParamDecl *param) {
+              return param->getName() == parsedWrtParams[i].getName();
+            });
+        // Parameter name must exist.
+        if (nameIter == params.end()) {
+          TC.diagnose(paramLoc, diag::diff_params_clause_param_name_unknown,
+                      parsedWrtParams[i].getName());
+          return nullptr;
+        }
+        // Parameter names must be specified in the original order.
+        unsigned index = std::distance(params.begin(), nameIter);
+        if ((int)index <= lastIndex) {
+          TC.diagnose(paramLoc,
+                      diag::diff_params_clause_params_not_original_order);
+          return nullptr;
+        }
+        builder.setParameter(index);
+        lastIndex = index;
+        break;
+      }
+      case ParsedAutoDiffParameter::Kind::Self: {
+        // 'self' is only applicable to instance methods.
+        if (!isInstanceMethod) {
+          TC.diagnose(paramLoc,
+                      diag::diff_params_clause_self_instance_method_only);
+          return nullptr;
+        }
+        // 'self' can only be the first in the list.
+        if (i > 0) {
+          TC.diagnose(paramLoc, diag::diff_params_clause_self_must_be_first);
+          return nullptr;
+        }
+        builder.setParameter(builder.size() - 1);
+        break;
+      }
+    }
+  }
+  return builder.build(TC.Context);
+}
+
+// SWIFT_ENABLE_TENSORFLOW
+// Checks if the given `AutoDiffParameterIndices` instance is valid for the
+// given function type in the given derivative generic environment and module
+// context. Returns true on error.
+// The parsed differentiation parameters and attribute location are used in
+// diagnostics.
+static bool checkDifferentiationParameters(
```

Comment:
It's possible to merge `checkDifferentiationParameters` into `computeDifferentiationParameters`, but that means synthesized attributes created with `AutoDiffParameterIndices` instead of `ArrayRef<ParsedAutoDiffParameter>` won't be verified.

Demo:
```cpp
  if (!checkedWrtParamIndices)
    checkedWrtParamIndices =
        computeDifferentiationParameters(TC, parsedWrtParams, original, // <- verification logic
                                         whereClauseGenEnv, attr->getAttrName(),
                                         attr->getLocation());
  if (!checkedWrtParamIndices) {
    attr->setInvalid();
    return;
  }

  // Uh oh, if `checkedWrtParamIndices` is already specified and not computed via
  // `computeDifferentiationParameters`, then it's not verified.
```

We could assume that all such synthesized attributes have valid `AutoDiffParameterIndices`, but it's safer to always validate.